google/resource_dataproc_cluster.go

```diff
@@ -0,0 +1,870 @@
+
+		wconfigs := d.Get("worker_config").([]interface{})
+		conf := wconfigs[0].(map[string]interface{})
+
+		desiredNumWorks := conf["preemptible_num_workers"].(int)
+		cluster.Config.SecondaryWorkerConfig = &dataproc.InstanceGroupConfig{
+			NumInstances: int64(desiredNumWorks),
+		}
+
+		if len(updMask) > 0 {
+			updMask = updMask + ","
+		}
+		updMask = updMask + "config.secondary_worker_config.num_instances"
+	}
+
+	patch.UpdateMask(updMask)
+
+	op, err := patch.Do()
+	if err != nil {
+		return err
+	}
+
+	// Wait until it's updated
+	waitErr := dataprocClusterOperationWait(config, op, "updating Dataproc cluster ", timeoutInMinutes, 2)
+	if waitErr != nil {
+		return waitErr
+	}
+
+	log.Printf("[INFO] Dataproc cluster %s has been updated ", d.Id())
+	return resourceDataprocClusterRead(d, meta)
+}
+
+func extractLastResourceFromUri(rUri string) string {
+	rUris := strings.Split(rUri, "/")
+	return rUris[len(rUris)-1]
+}
+
+func resourceDataprocClusterRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region := d.Get("region").(string)
+	clusterName := d.Get("name").(string)
+
+	cluster, err := config.clientDataproc.Projects.Regions.Clusters.Get(
+		project, region, clusterName).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Dataproc Cluster %q", clusterName))
+	}
+
+	d.Set("labels", cluster.Labels)
+	d.Set("name", cluster.ClusterName)
+	d.Set("bucket", cluster.Config.ConfigBucket)
+
+	extracted := false
+	if len(cluster.Config.InitializationActions) > 0 {
+		actions := []string{}
+		for _, v := range cluster.Config.InitializationActions {
+			actions = append(actions, v.ExecutableFile)
+
+			if !extracted && len(v.ExecutionTimeout) > 0 {
+				tsec, err := extractInitTimeout(v.ExecutionTimeout)
+				if err != nil {
+					return err
+				}
+				d.Set("initialization_action_timeout_sec", tsec)
+				extracted = true
+			}
+		}
+		d.Set("initialization_actions", actions)
+	}
+
+	if cluster.Config.GceClusterConfig != nil {
+		d.Set("zone", extractLastResourceFromUri(cluster.Config.GceClusterConfig.ZoneUri))
+		d.Set("network", extractLastResourceFromUri(cluster.Config.GceClusterConfig.NetworkUri))
+		d.Set("subnet", extractLastResourceFromUri(cluster.Config.GceClusterConfig.SubnetworkUri))
+		d.Set("tags", cluster.Config.GceClusterConfig.Tags)
+		d.Set("metadata", cluster.Config.GceClusterConfig.Metadata)
+		d.Set("service_account", cluster.Config.GceClusterConfig.ServiceAccount)
+		if len(cluster.Config.GceClusterConfig.ServiceAccountScopes) > 0 {
+			sort.Strings(cluster.Config.GceClusterConfig.ServiceAccountScopes)
+			d.Set("service_account_scopes", cluster.Config.GceClusterConfig.ServiceAccountScopes)
+		}
+
+	}
+
+	if cluster.Config.SoftwareConfig != nil {
+		d.Set("image_version", cluster.Config.SoftwareConfig.ImageVersion)
+		//We only want our overriden values here for now
```

Comment:
Because GCP returns ALL properties, both overridden by user and the default ones. If it we set the values here, the user would get a diff on a plan the next time as it would suddenly now include the 100's of default options.